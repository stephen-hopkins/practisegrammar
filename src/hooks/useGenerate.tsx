import { useMutation } from "@tanstack/react-query";
import { Translation } from "@/src/models/translation";
import { Concept } from "@/src/models/constants";

type useGenerateInput = {
  concept: Concept;
  word: string;
};

function useGenerate() {
  return useMutation(async ({ concept, word }: useGenerateInput) => {
    const response = await fetch(`/api/sentences/Russian/${concept}`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        word,
      }),
    });
    const trans = await response.json();
    return trans as Translation;
  });
}

export default useGenerate;
