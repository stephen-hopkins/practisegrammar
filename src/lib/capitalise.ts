export function capitalise(word: string) {
  return word.charAt(0).toLocaleLowerCase() + word.slice(1);
}