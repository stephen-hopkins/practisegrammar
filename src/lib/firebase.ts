import { initializeApp } from "firebase/app";
import { Concept, Language } from "@/src/models/constants";
import { addDoc, collection, CollectionReference, getDocs, getFirestore, query, where } from "@firebase/firestore";
import { Sentence, SentenceInfo } from "@/src/models/translation";

const firebaseConfig = {
  apiKey: "AIzaSyCXVeg1hzBtWVW2FK9_cz7r7gygHIH8E2k",
  authDomain: "practisegrammar.firebaseapp.com",
  projectId: "practisegrammar",
  storageBucket: "practisegrammar.appspot.com",
  messagingSenderId: "666711626320",
  appId: "1:666711626320:web:410a88534639301d4b7755",
  measurementId: "G-YCLPJ0Z093",
};

function getDb() {
  // const analytics = getAnalytics(app);
  const app = initializeApp(firebaseConfig);
  const db = getFirestore(app);

  return {
    sentences: collection(db, "sentences") as CollectionReference<SentenceInfo>,
  };
}

export async function getSentence(language: Language, concept: Concept, word: string) {
  const { sentences } = getDb();
  const q = query(
    sentences,
    where("language", "==", language),
    where("concept", "==", concept),
    where("word", "==", word)
  );
  const querySnapshot = await getDocs(q);
  if (!querySnapshot.empty) {
    return querySnapshot.docs[0].data();
  }
}

export async function addSentence(sentence: Sentence) {
  const { translations, ...metaData } = sentence;
  const { sentences } = getDb();
  //const doc = await addDoc(sentences, metaData, "translations", translations);
  //await addDoc(collection(sentences, doc.id, "translations"), translations);
}
