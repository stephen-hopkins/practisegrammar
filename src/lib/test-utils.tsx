import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { RenderOptions, render } from "@testing-library/react";
import "@testing-library/jest-dom";
import { MemoryRouterProvider } from "next-router-mock/MemoryRouterProvider";

jest.mock("next/navigation", () => require("next-router-mock"));

const queryClient = new QueryClient();
const AllProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <QueryClientProvider client={queryClient}>
      <MemoryRouterProvider url="/">{children}</MemoryRouterProvider>
    </QueryClientProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
