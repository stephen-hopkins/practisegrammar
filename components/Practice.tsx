import React, {useState} from 'react';
import {InputText} from "primereact/inputtext";
import {classNames} from "primereact/utils";
import {Button} from "primereact/button";
import useGenerate from "@/hooks/useGenerate";

function Practice() {

  const [word, setWord] = useState('');
  const [showTrans, setShowTrans] = useState(false);
  const generate = useGenerate();

  const onSubmit = () => {
    setShowTrans(false);
    generate.mutate(word);
  }

  const english = generate.isSuccess && generate.data.english;
  const russian = generate.isSuccess && generate.data.russian;

  return (
    <div className="flex flex-column align-items-center">
      <div>
        <InputText value={word} onChange={e => setWord(e.target.value)} />
        <Button className="ml-4" onClick={onSubmit}>Generate</Button>
      </div>
      <div className="mt-6 text-4xl">{english}</div>
      <Button className="mt-4" onClick={() => setShowTrans(true)}>Reveal</Button>
      <div className={classNames("mt-4 text-4xl", {hidden: !showTrans})}>{russian}</div>
    </div>
  );
}

export default Practice;