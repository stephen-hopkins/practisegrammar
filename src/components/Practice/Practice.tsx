import React, { useState } from "react";
import { InputText } from "primereact/inputtext";
import { Button } from "primereact/button";
import { Concept, Concepts } from "@/src/models/constants";
import { Dropdown } from "primereact/dropdown";
import { ProgressSpinner } from "primereact/progressspinner";
import SpeechBubble from "../SpeechBubble/SpeechBubble";
import useGenerate from "@/src/hooks/useGenerate";

function Practice() {
  const [concept, setConcept] = useState(undefined as Concept | undefined);
  const [word, setWord] = useState("");
  const [showTrans, setShowTrans] = useState(false);
  const generate = useGenerate();

  const onSubmit = () => {
    if (word && concept) {
      setShowTrans(false);
      generate.mutate({ concept, word });
    }
  };

  const english = generate.isSuccess && generate.data.english;
  const russian = generate.isSuccess && generate.data.russian;

  const toggleShowTrans = () => {
    setShowTrans(!showTrans);
  };

  return (
    <div className="grid">
      <p className="col-12 m-0">Which case would you like to practice?</p>
      <div className="col-12 flex justify-content-end">
        <Dropdown
          placeholder="Select a case"
          className="w-auto"
          value={concept}
          onChange={(e) => setConcept(e.target.value)}
          options={[...Concepts]}
          data-testid="concept-dropdown"
        />
      </div>
      <p className="col-12 ml-auto m-0">Which word should be included?</p>
      <div className="col-12 flex justify-content-end">
        <InputText className="" value={word} onChange={(e) => setWord(e.target.value)} placeholder="word to practice" />
      </div>
      <div className="col-12 flex justify-content-center">
        <Button className="mt-2" onClick={onSubmit}>
          Generate
        </Button>
      </div>
      {generate.isLoading && <ProgressSpinner className="mt-4" />}
      <SpeechBubble
        className="col-12"
        text={english ? english : ""}
        type="original"
        show={!!english}
        onClick={toggleShowTrans}
      />
      <SpeechBubble
        className="col-12"
        text={russian ? russian : ""}
        type="translated"
        show={showTrans}
        onClick={toggleShowTrans}
      />
    </div>
  );
}

export default Practice;
