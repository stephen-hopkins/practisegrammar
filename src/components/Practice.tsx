import React, { useMemo, useState } from "react";
import { InputText } from "primereact/inputtext";
import { classNames } from "primereact/utils";
import { Button } from "primereact/button";
import useGenerate from "@/src/hooks/useGenerate";
import { Concept, Concepts } from "@/src/models/constants";
import { Dropdown } from "primereact/dropdown";
import { ProgressSpinner } from "primereact/progressspinner";
import { capitalise } from "@/src/lib/capitalise";
import SpeechBubble from "./SpeechBubble";

function Practice() {
  const [concept, setConcept] = useState("Accusative" as Concept);
  const [word, setWord] = useState("");
  const [showTrans, setShowTrans] = useState(false);
  const generate = useGenerate();

  const onSubmit = () => {
    setShowTrans(false);
    generate.mutate({ concept, word });
  };

  const english = generate.isSuccess && generate.data.english;
  const russian = generate.isSuccess && generate.data.russian;

  const conceptOptions = useMemo(() => {
    return Concepts.map((c) => {
      return {
        label: capitalise(c),
        value: c,
      };
    });
  }, []);

  return (
    <div className="flex flex-column align-items-center">
      <Dropdown value={concept} onChange={(e) => setConcept(e.target.value)} options={conceptOptions} />
      <InputText className="mt-4" value={word} onChange={(e) => setWord(e.target.value)} />
      <Button className="mt-4" onClick={onSubmit}>
        Generate
      </Button>
      {generate.isLoading && <ProgressSpinner className="mt-4" />}
      <SpeechBubble text={english ? english : ""} type="original" show={!!english} />
      <Button className={classNames("mt-4", { hidden: !generate.isSuccess })} onClick={() => setShowTrans(true)}>
        Reveal
      </Button>
      <SpeechBubble text={russian ? russian : ""} type="translated" show={showTrans} />
    </div>
  );
}

export default Practice;
