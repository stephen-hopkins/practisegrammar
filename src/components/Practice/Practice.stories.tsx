import { Meta, StoryObj } from "@storybook/react";
import Practice from "./Practice";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

const queryClient = new QueryClient();
const meta: Meta<typeof Practice> = {
  title: "Practice",
  component: Practice,
  decorators: [
    (Story) => (
      <QueryClientProvider client={queryClient}>
        <Story />
      </QueryClientProvider>
    ),
  ],
};

export default meta;

type Story = StoryObj<typeof Practice>;

export const Default: Story = {
  args: {},
};
