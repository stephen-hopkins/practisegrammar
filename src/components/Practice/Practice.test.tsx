import { render, screen } from "@/src/lib/test-utils";
import Practice from "./Practice";
import userEvent from "@testing-library/user-event";
import useGenerate from "@/src/hooks/useGenerate";

const mutateSpy = jest.fn();
jest.mock("@/src/hooks/useGenerate", () => {
  return {
    __esModule: true,
    default: jest.fn(() => {
      return {
        mutate: mutateSpy,
        isLoading: false,
        isSuccess: true,
        data: {
          english: "Example english sentence",
          russian: "Example russian sentence",
        },
      };
    }),
  };
});

describe("Practice", () => {
  it("renders", () => {
    render(<Practice />);
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<Practice />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("calls onSubmit with concept and word when generate is clicked", async () => {
    const user = userEvent.setup();
    const { getByPlaceholderText, getByTestId, getByText } = render(<Practice />);
    const conceptDropdown = getByTestId("concept-dropdown");
    const wordInput = getByPlaceholderText("word to practice");
    const generateButton = getByText("Generate");

    await user.click(conceptDropdown);
    await user.click(getByText("accusative"));
    await user.type(wordInput, "hello");
    await user.click(generateButton);

    expect(useGenerate().mutate).toHaveBeenCalledWith({
      concept: "accusative",
      word: "hello",
    });
  });

  it("displays english sentence", async () => {
    const user = userEvent.setup();
    render(<Practice />);

    const generateButton = screen.getByText("Generate");
    await user.click(generateButton);

    expect(screen.getByText("Example english sentence")).toBeInTheDocument();
    const russianBubble = screen.queryByText('Example russian sentence')
    expect(russianBubble).toBeNull();
  });
});
