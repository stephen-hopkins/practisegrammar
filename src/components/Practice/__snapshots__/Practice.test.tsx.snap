// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Practice matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid"
  >
    <div
      class="col-12 flex justify-content-center"
    >
      <p
        class="m-0"
      >
        Which case would you like to practice?
      </p>
    </div>
    <div
      class="col-12 flex justify-content-center"
    >
      <div
        class="p-dropdown p-component p-inputwrapper w-auto"
        data-testid="concept-dropdown"
      >
        <div
          class="p-hidden-accessible"
        >
          <input
            aria-haspopup="listbox"
            readonly=""
            type="text"
          />
        </div>
        <div
          class="p-hidden-accessible p-dropdown-hidden-select"
        >
          <select
            aria-hidden="true"
            tabindex="-1"
          >
            <option
              selected=""
              value=""
            >
              Select a case
            </option>
          </select>
        </div>
        <span
          class="p-dropdown-label p-inputtext p-placeholder"
        >
          Select a case
        </span>
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-label="Select a case"
          class="p-dropdown-trigger"
          role="button"
        >
          <span
            class="p-dropdown-trigger-icon p-clickable pi pi-chevron-down"
          />
        </div>
      </div>
    </div>
    <div
      class="col-12 flex justify-content-center"
    >
      <p
        class="m-0"
      >
        Which word should be included?
      </p>
    </div>
    <div
      class="col-12 flex justify-content-center"
    >
      <input
        class="p-inputtext p-component"
        placeholder="word to practice"
        value=""
      />
    </div>
    <div
      class="col-12 flex justify-content-center"
    >
      <button
        class="p-button p-component mt-2"
      >
        Generate
      </button>
    </div>
    <div
      class="col-12 flex justify-content-center"
    >
      <div
        class="right bubble"
      >
        <div
          class="text_wrapper"
        >
          <div
            class="text"
          >
            Example english sentence
          </div>
        </div>
      </div>
    </div>
    <div
      class="col-12 flex justify-content-center"
    />
  </div>
</DocumentFragment>
`;
